### PR TITLE
tech :: Convert vue components to TypeScript

### PR DESCRIPTION
**Pull request by davinov**
So we can benefit from type checking inside vue components.

Docs to learn how to use these decorators and class syntax:
- https://vuejs.org/v2/guide/typescript.html
- https://github.com/vuejs/vue-class-component
- https://github.com/kaorun343/vue-property-decorator

Note: storybook now uses a specific bundle, also built with the same rollup configuration.
It's therefore needed to export any component to showcase in `.storybook/bundle.ts`.

### CODE REVIEW

File: src/components/DomainSelector.vue [hunk 1]
```diff
@@ -16,22 +16,31 @@
...
+
+@Component({
+  name: 'domain-selector'
+})
+export default class DomainSelector extends Vue {
+  @Prop({
+    default: () => [],
+    type: Array
+  })
+  readonly domainsList!: Array<string>
```
Comment (davinov) [inline src/components/DomainSelector.vue]: So there is two type declaration here:
- the one in the decorator @Prop is meant for vue, and will be enforced at _run time_
- the other directly after the prop name is meant for typescript, and will be enforced at _transpile time_

File: .storybook/bundle.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+/*
+  This file creates the bundle needed by storybook to import the UI components.
+  All components available in the Storybook must be exported here.
+*/
+
+import Cell from '../src/components/dataset-viewer/Cell.vue'
```
Comment (adimascio) [inline .storybook/bundle.ts]: This component seems to be part of another PR, you should remove the import here.

File: .storybook/rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import mainConfig from '../rollup.config'
+
+let storybookComponentsPackageConfig = Object.assign({}, mainConfig)
```
Comment (adimascio) [inline .storybook/rollup.config.js]: Use `const` instead of `let` ? I don't see any new assignment to `storybookComponentsPackageConfig`.

Also, this is a just a matter of style but I like to use the spread operator for object assignments. Here, `Object.assign{{}, mainconfig)` could become `{...mainConfig}`. What do you think?

File: .storybook/rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+storybookComponentsPackageConfig.input = '.storybook/bundle.ts'
+storybookComponentsPackageConfig.output = [
+  {
+    file: 'dist/storybook/components.js',
+    format: 'esm'
+  }
+]
+
+export default storybookComponentsPackageConfig
+
```
Comment (adimascio) [inline .storybook/rollup.config.js]: Actually, the whole assignment could become:

```javascript

export default {
    ...mainConfig,
    input: '.storybook/bundle.ts',
    output: [
        {
            file: 'dist/storybook/components.js',
            format: 'esm'
        }
    ],
};
```

Comment (adimascio) [reply .storybook/rollup.config.js]: again, this is purely a matter of stlye :)

File: src/components/DomainSelector.vue (hunk 1 above)
Comment (adimascio) [reply src/components/DomainSelector.vue]: :+1:

File: src/components/DomainSelector.vue [hunk 5]
```diff
@@ -16,22 +16,31 @@
...
+<script lang="ts">
+import Vue from 'vue'
+import { Component, Prop } from 'vue-property-decorator'
+
+@Component({
+  name: 'domain-selector'
+})
+export default class DomainSelector extends Vue {
+  @Prop({
+    default: () => [],
```
Comment (adimascio) [inline src/components/DomainSelector.vue]: Why not `deafult: []` directly? Is it a matter of mutability and default being evaluated once and for all?

File: src/components/DomainSelector.vue [hunk 6]
```diff
@@ -16,22 +16,31 @@
...
+    type: Array
+  })
+  readonly domainsList!: Array<string>
+
+  @Prop()
+  readonly selectedDomain!: string | undefined
```
Comment (Dalanir) [inline src/components/DomainSelector.vue]: the bang operator is not supposed to exclude `undefined` value ?

File: src/components/DomainSelector.vue (hunk 1 above)
Comment (adimascio) [reply src/components/DomainSelector.vue]: :+1:

File: src/components/DomainSelector.vue (hunk 6 above)
Comment (adimascio) [reply src/components/DomainSelector.vue]: The `!` tells the typescript compiler not to crash if the attribute ist not initialized in constructor. In case of props, I guess it's a bit trickier. Probably related to https://github.com/vuejs/vue-class-component/issues/221

File: .storybook/rollup.config.js (hunk 4 above)
Comment (davinov) [reply .storybook/rollup.config.js]: And clarity, I strongly agree ;)

File: .storybook/rollup.config.js (hunk 3 above)
Comment (davinov) [reply .storybook/rollup.config.js]: I'm not that sure here: I know that `const` only forbids the identifier to be reassigned, yet it passes the message that is object is quite "constant", i.e. immutable.

The lines just below reassigned properties of this object, so I was not that inclined to use `const`.

File: src/components/DomainSelector.vue (hunk 5 above)
Comment (davinov) [reply src/components/DomainSelector.vue]: Yes, otherwise it's shared among every instance of this component: https://vuejs.org/v2/api/#props

File: .storybook/bundle.ts (hunk 2 above)
Comment (davinov) [reply .storybook/bundle.ts]: Oops, it's effectively in the next one ;)

File: .storybook/rollup.config.js (hunk 3 above)
Comment (adimascio) [reply .storybook/rollup.config.js]: I understand but still, I disagree :-). My preference would be to stick with the language semantic and maybe even enforce the eslint `prefer-const: "error"` rule. Any objection?

File: src/components/DomainSelector.vue (hunk 5 above)
Comment (adimascio) [reply src/components/DomainSelector.vue]: :+1:

File: src/components/DomainSelector.vue (hunk 6 above)
Comment (davinov) [reply src/components/DomainSelector.vue]: Yep, @adimascio is right: as vue will assign itself the prop depending on what the parent component bound to it, we do not pass an initiaa value

vue-class-component suggest to use the `!` to still get type hinting on props

**Comment by davinov** [comment]
@adimascio Thanks for the feedback, I've made the changes and rebased for clarity
